docs/decouple-application-sync-user-using-impersonation

```diff
@@ -0,0 +1,269 @@
+
+### Assumptions
+
+- Namespaces are pre-populated with one or more ServiceAccounts that define the permissions for each application team (or application).
+- Users prefer to control access to k8s resources through kubernetes RBAC constructs instead of Argo specific constructs.
+- Each tenant is generally given access to a specific namespace along with a service account, role or cluster role and role binding to control access to that namespace.
+
+
+### Goals
+- Applications may only impersonate ServiceAccounts that live in the same namespace as the application itself. Application teams can then choose the appropriate ServiceAccount to be used for syncing each application themselves, without being able to elevate their privileges.
+- If impersonation feature is enabled, and no service account name is provided in the Application CR, then it should use a default service account setting specified in the AppProject and if not specified use the namespace's default service account.
+- Access restrictions implemented through properties in AppProject (if done) must have the existing behavior. From a security standpoint, any restrictions that were available before switching to a service account based approach should continue to exist even when the impersonation feature is enabled.
+
+### Non-Goals
+
+What is out of scope for this proposal? Listing non-goals helps to focus discussion and make
+progress.
+
+## Proposal
+
+- Enable users to specify a `serviceAccountName` in an Application CR. The service account should exist in the same namespace as that of the Application CR.
+
+```
+apiVersion: argoproj.io/v1alpha1
+kind: Application
+metadata:
+  name: guestbook
+  namespace: argocd
+spec:
+  project: default
+  source:
+    repoURL: https://github.com/argoproj/argocd-example-apps.git
+    targetRevision: HEAD
+    path: guestbook
+  destination:
+    server: https://kubernetes.default.svc
+    namespace: guestbook
+  # This application would use the below service account to impersonate during the sync operation
+  serviceAccountName: guestbook-deployer
+```
+
+Enable ArgoCD Admin to specify a `defaultServiceAccountName` in `AppProjects` CR. This would be used for all Argo Applications associated with that AppProject if not specifically overridden in the Application CR.
+
+```
+apiVersion: argoproj.io/v1alpha1
+kind: AppProject
+metadata:
+  name: my-project
+  namespace: argocd
+  finalizers:
+    - resources-finalizer.argocd.argoproj.io
+spec:
+  description: Example Project
+  # Allow manifests to deploy from any Git repos
+  sourceRepos:
+  - '*'
+  destinations:
+  - namespace: guestbook
+    server: https://kubernetes.default.svc
```

Comment:
Could we not have the `defaultServiceAccountName` in the destinations list instead? Then we'd have the possibility of doing something like below:

```yaml
destinations:
- namespace: guestbook
  server: https://kubernetes.default.svc
  defaultServiceAccountName: foo
- namespace: guestbook-2
  server: https://kubernetes.default.svc
  defaultServiceAccountName: bar
- namespace: other-namespaces*
  server: https://kubernetes.default.svc
  defaultServiceAccountName: bat
```

Reply:
With this there shouldn't be a need to allow users to set `serviceAccountName` directly in the `Application` CRD. This would mean that this can be more tightly controlled by the Argo CD admins.

Reply:
This could perhaps be extended by adding `allowedServiceAccounts`, so that a destinations block could look as below:

```yaml
destinations:
- namespace: guestbook
  server: https://kubernetes.default.svc
  defaultServiceAccountName: foo
  allowedServiceAccounts: 
  - foo
  - bar
  - bat
- namespace: guestbook-2
  server: https://kubernetes.default.svc
  defaultServiceAccountName: bar
  allowedServiceAccounts: 
  - foo
  - bar
  - bat
- namespace: other-namespaces*
  server: https://kubernetes.default.svc
  defaultServiceAccountName: bat
  allowedServiceAccounts: 
  - foo
  - bar
  - bat
```

Reply:
Thanks @blakepettersson for all your review comments and the doc fixes (which I missed) :-) 
We too, had similar internal discussions of restricting the service account config only at the `AppProject`. Allowing it at the `Application` level was to give more flexibility to users to handle scenarios where there is a need to have 1 service account per Application. But, since we need to restrict the service account specified in the `Application`, it probably makes sense to have all the control in the `AppProject` itself. I like the idea of having it part of the `destinations` array as it allows us to specify 1 per destination target. Let me see what others think about this approach and then make the change in the proposal.